flit/upload.py

```diff
@@ -160,21 +174,44 @@ def get_password(repo, prefer_env):
     if sys.stdin.isatty():
         pw = None
         while not pw:
-            print('Server  :', repo['url'])
-            print('Username:', repo['username'])
+            print('Server  :', repo.url)
+            print('Username:', repo.username)
             pw = getpass.getpass('Password: ')
     else:
         raise Exception("Could not find password for upload.")
 
     if keyring is not None:
         try:
-            keyring.set_password(repo['url'], repo['username'], pw)
+            keyring.set_password(repo.url, repo.username, pw)
             log.info("Stored password with keyring")
         except keyring.errors.KeyringError as e:
             log.warning("Could not store password in keyring (%s)", e)
 
     return pw
 
+
+def find_token(repo: RepoDetails, project_name: str):
+    # https://packaging.python.org/en/latest/specifications/name-normalization/
+    project_name = re.sub(r"[-_.]+", "-", project_name).lower()
+    candidate_keys = [f"pypi_token:project:{project_name}"]
+
+    if repo.username is not None:
+        candidate_keys.append(f"pypi_token:user:{repo.username}")
+
+    try:
+        import keyring, keyring.errors
+    except ImportError:  # pragma: no cover
```

Comment:
Do you want to catch only `ModuleNotFoundError` and not all ImportErrors ?

Reply:
I think I prefer the broader `ImportError` - if importing keyring fails for any reason, treat it as not available. But in practice I don't think it's a big difference, as since [almost 3 years ago](https://github.com/jaraco/keyring/pull/480), keyring doesn't load its backends on import, so there's not much happening on import that's likely to fail.

Reply:
I certainly tend to forget the difference, so the reminder is welcome. :slightly_smiling_face: